File: lua/application.lua
local lgi = require("lgi")
local Gtk = lgi.require("Gtk", "3.0")

-- GtkApplication: Provides an application interface

local appID = "io.github.Miqueas.GTK-Examples.Lua.Gtk3.Application"
local app = Gtk.Application({ application_id = appID })

-- I recommend to check out this links for more information:
-- https://wiki.gnome.org/HowDoI/GtkApplication
-- https://developer.gnome.org/gio/stable/GApplication.html
-- https://developer.gnome.org/gtk3/stable/GtkApplication.html
function app:on_startup()
  -- The ::startup signal is called only once, so it's recommended to create/prepare
  -- all the GUI here and then present it on the ::activate signal
  print("Application::startup called")
end

function app:on_activate()
  -- The ::activate signal can be triggered indefinitely, so please try to not do heavy
  -- things here. Check out the doc links above for more details
  print("Application::activate called")
end

return app:run(arg)

File: cpp/gtk3/oop/application.cpp
#include <print>
#include <gtkmm.h>

static const Glib::ustring APP_ID = "io.github.Miqueas.GTK-Examples.Cpp.Gtk3.OOP.Application";

class App : public Gtk::Application {
  public:
    static Glib::RefPtr<App> create() {
      return Glib::RefPtr<App>(new App());
    }

  protected:
    App() : Gtk::Application(APP_ID) {}

    void on_activate() override {
      std::println("::activate");
    }

    void on_startup() override {
      // Ensures chain-up
      Gtk::Application::on_startup();
      std::println("::startup");
    }
};

int main(int argc, char** argv) {
  return App::create()->run(argc, argv);
}